### ISSUE TITLE
Add `glue` primitive

### ISSUE THREAD

**Issue by MatthewDaggitt** [vehicle-lang/vehicle#507]
The function `glue : A -> A` should be semantically the identity function, but when `glue x` is normalised results in a `Glue` constructor that contains both `x` and the unnormalised form of `x`. In this way, we can remember both the normalised and unnormalised form of expressions when required. For example in the loss function backend when translating vector operations to `numpy` operations or the verification backend when trying to solve equations at the vector level.

**Comment by wenkokke** [comment]
Glue should not be a language primitive, but rather the default behavior when normalizing any top-level definition. We don't need to store an unnormalised term, but only a definition name and a spine.

**Comment by wenkokke** [comment]
We decided against this because gluing seems incompatible with branch elimination.

If we represent glue as `Glue (ident, spine) result` where the first argument is the glued function with its arguments and the second is the normal form, we'd like to maintain that `eval (App ident spine) == result`.

We certainly cannot lift ifs out of the definition of ident without changing the semantics of the glue node... and introducing a fresh declaration with an if-lifted interpretation would make the glue itself useless.

However, if we want to lift ifs out of the spine, we're lifting them past the glue node, but the context around the ifs in the spine will be different from the context around the ifs in the result, so we'd have to store these contexts on the glue node as well, which means that a glue node would end up storing not only the expression below it, but its entire context up to the nearest enclosing boolean expression, as that's where we lift ifs... which would make the semantics of glue extremely intricate.

**Comment by wenkokke** [comment]
TL;DR gluing stores two alternative local term representations, but if lifting is a global transformation, which makes the two fundamentally incompatible.
